cluster-agent/controllers/argoproj.io/namespace_reconciler.go

```diff
@@ -0,0 +1,321 @@
+						Resource_type: db.OperationResourceType_Application,
+					}
+
+					_, _, err = eventlooptypes.CreateOperation(ctx, false, dbOperationInput,
+						specialClusterUser.Clusteruser_id, cache.GetGitOpsEngineSingleInstanceNamespace(), dbQueries, client, log)
+					if err != nil {
+						log.Error(err, "Namespace Reconciler is unable to create operation: "+dbOperationInput.ShortString())
+					}
+					log.Info("Operation is created to recreateArgoCD  Application " + applicationRowFromDB.Application_id)
+					continue
+				} else {
+					log.Error(err, "Error occurred in Namespace Reconciler while fetching application from cluster: "+applicationRowFromDB.Application_id)
+					continue
+				}
+			}
+
+			// At this point we have the applications from ArgoCD and DB, now compare them to check if they are not in Sync.
+			if compareApplications(applicationFromArgoCD, applicationFromDB, log) {
+				log.V(sharedutil.LogLevel_Debug).Info("Argo application is in Sync with DB, Application:" + applicationRowFromDB.Application_id)
+				continue
+			} else {
+				log.Info("Argo application is not in Sync with DB, updating Argo CD App. Application:" + applicationRowFromDB.Application_id)
+			}
+
+			// At this point application from ArgoCD and DB are not in Sync, so need to update ArgoCD application according to DB entry
+
+			// ArgoCD application and DB entry are not in Sync,
+			// ArgoCD should use the state of resources present in the database should
+			// Create Operation to inform ArgoCD to get in Sync with database entry.
+			dbOperationInput := db.Operation{
+				Instance_id:   applicationRowFromDB.Engine_instance_inst_id,
+				Resource_id:   applicationRowFromDB.Application_id,
+				Resource_type: db.OperationResourceType_Application,
+			}
+
+			_, _, err = eventlooptypes.CreateOperation(ctx, false, dbOperationInput,
+				specialClusterUser.Clusteruser_id, cache.GetGitOpsEngineSingleInstanceNamespace(), dbQueries, client, log)
+			if err != nil {
+				log.Error(err, "Namespace Reconciler is unable to create operation: "+dbOperationInput.ShortString())
+				continue
+			}
+
+			log.Info("Namespace Reconcile processed application: " + applicationRowFromDB.Application_id)
+		}
+
+		// Skip processed entries in next iteration
+		offSet += appRowBatchSize
+	}
+
+	// Start a goroutine, because DeleteArgoCDApplication() function from cluster-agent/controllers may take some time to delete application.
+	go deleteOrphanedApplications(argoApplications, processedApplicationIds, ctx, client, log)
+
+	log.Info(fmt.Sprintf("Namespace Reconciler finished an iteration at %s. "+
+		"Next iteration will be triggered after %v Minutes", time.Now().String(), namespaceReconcilerInterval))
+}
+
+// compareApplications compares Application objects, since both objects are of different types we can not use == operator for comparison.
+func compareApplications(applicationFromArgoCD appv1.Application, applicationFromDB fauxargocd.FauxApplication, log logr.Logger) bool {
```

Comment:
I had similar implementation but then as @jgwest noticed, this logic will issue an update to the object per different field. In other words, if there are e.g. 3 differences, it will correct the object 3 times, aka it will send 3 API calls. Ideally we would like to minimize the API calls. So if we can send one call and update it all at once it would be preferable. 

Take a look at my implementation for a very similar function: https://github.com/redhat-appstudio/managed-gitops/pull/168/files#diff-a387a638ba7c6a4d686c64d88c5b9c9a28a7b289ed665fed4baed16c99c6ab37R246-R250

Reply:
Yes, there is no harm in doing this, but here it is not exactly same case. In compareApplications() if any of the field is different, it will return from there only and will not compare any other field. After this a new Operation will be created and all out of sync fields will be updated at once. Means if there are 3 different fields still only 1 Operation will be created, which will update all 3 fields in by making 1 API call.
https://github.com/redhat-appstudio/managed-gitops/blob/f4121710a35ef41d156da68bbb54c22748bc6cd5/cluster-agent/controllers/argoproj.io/namespace_reconciler.go#L152

Reply:
O I've got it now ;) One small caveat in that case you are printing only the first log. So if there are more than one diffs you miss the rest.